retrieval/discovery/serverset.go

```diff
@@ -0,0 +1,359 @@
+// Copyright 2015 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package discovery
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"regexp"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/prometheus/log"
+	"github.com/samuel/go-zookeeper/zk"
+
+	clientmodel "github.com/prometheus/client_golang/model"
+
+	"github.com/prometheus/prometheus/config"
+)
+
+const (
+	serversetSourcePrefix = "serverset"
+
+	serversetNodePrefix = "member_"
+
+	serversetLabelPrefix         = clientmodel.MetaLabelPrefix + "serverset_"
+	serversetStatusLabel         = serversetLabelPrefix + "status"
+	serversetPathLabel           = serversetLabelPrefix + "path"
+	serversetEndpointLabelPrefix = serversetLabelPrefix + "endpoint"
+)
+
+var (
+	invalidLabelCharRE = regexp.MustCompile(`[^a-zA-Z0-9_]`)
+)
+
+type serversetMember struct {
+	ServiceEndpoint     serversetEndpoint
+	AdditionalEndpoints map[string]serversetEndpoint
+	Status              string `json:"status"`
+}
+
+type serversetEndpoint struct {
+	Host string
+	Port int
+}
+
+type ZookeeperLogger struct {
+}
+
+// Implements zk.Logger
+func (zl ZookeeperLogger) Printf(s string, i ...interface{}) {
+	log.Infof(s, i...)
+}
+
+// ServersetDiscovery retrieves target information from a Serverset server
+// and updates them via watches.
+type ServersetDiscovery struct {
+	conf      *config.ServersetSDConfig
+	conn      *zk.Conn
+	mu        sync.RWMutex
+	sources   map[string]struct{}
+	updates   chan ZookeeperTreeCacheEvent
+	runDone   chan struct{}
+	treeCache *ZookeeperTreeCache
+}
+
+// NewServersetDiscovery returns a new ServersetDiscovery for the given config.
+func NewServersetDiscovery(conf *config.ServersetSDConfig) *ServersetDiscovery {
+	conn, _, err := zk.Connect(conf.Servers, time.Duration(conf.Timeout))
+	conn.SetLogger(ZookeeperLogger{})
+	if err != nil {
+		return nil
+	}
+	updates := make(chan ZookeeperTreeCacheEvent, 20000)
+	sd := &ServersetDiscovery{
+		conf:      conf,
+		conn:      conn,
+		updates:   updates,
+		sources:   make(map[string]struct{}),
```

Comment:
We have the "no `make()` unless necessary" style guideline, so `map[string]struct{}{}`.